controllers/containerizedworkload_controller.go

```diff
@@ -0,0 +1,156 @@
+)
+
+const (
+	oamReconcileWait = 30 * time.Second
+)
+
+// Reconcile error strings.
+const (
+	errRenderWorkload  = "cannot render workload"
+	errUpdateStatus    = "cannot apply status"
+	errApplyDeployment = "cannot apply the deployment"
+	errApplyService    = "cannot apply the service"
+	errGCDeployment    = "cannot clean up stale deployments"
+	errScaleDeployment = "cannot scale the deployment"
+)
+
+// ContainerizedWorkloadReconciler reconciles a ContainerizedWorkload object
+type ContainerizedWorkloadReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=core.oam.dev,resources=containerizedworkloads,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core.oam.dev,resources=containerizedworkloads/status,verbs=get;update;patch
+func (r *ContainerizedWorkloadReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+	log := r.Log.WithValues("containerizedworkload", req.NamespacedName)
+	log.Info("Reconcile container workload")
+
+	var workload oamv1alpha2.ContainerizedWorkload
+	if err := r.Get(ctx, req.NamespacedName, &workload); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	log.Info("Get the workload")
+
+	deploy, err := r.renderWorkload(ctx, &workload)
+	if err != nil {
+		workload.Status.SetConditions(cpv1alpha1.ReconcileError(errors.Wrap(err, errRenderWorkload)))
+		log.Error(err, "Failed to render a deployment")
+		return reconcile.Result{RequeueAfter: oamReconcileWait}, errors.Wrap(r.Status().Update(ctx, &workload),
+			errUpdateStatus)
+	}
+	log.Info("Successfully rendered a deployment", "deployment", deploy.Name)
+
+	// server side apply, only the fields we set are touched
+	applyOpts := []client.PatchOption{client.ForceOwnership, client.FieldOwner(workload.ObjectMeta.Name)}
+	if err := r.Patch(ctx, deploy, client.Apply, applyOpts...); err != nil {
+		workload.Status.SetConditions(cpv1alpha1.ReconcileError(errors.Wrap(err, errApplyDeployment)))
+		log.Error(err, "Failed to apply to a deployment")
+		return reconcile.Result{RequeueAfter: oamReconcileWait}, errors.Wrap(r.Status().Update(ctx, &workload),
+			errUpdateStatus)
+	}
+	log.Info("Successfully applied a deployment", "UID", deploy.UID)
+
+	// record the new deployment
+	if len(workload.Status.Resources) == 0 {
+		workload.Status.Resources = make(map[types.UID]oamv1alpha2.ResourceReference)
+	}
+	workload.Status.Resources[deploy.UID] = oamv1alpha2.ResourceReference{
+		APIVersion: deploy.APIVersion,
+		Kind:       deploy.Kind,
+		Name:       deploy.Name,
+		UID:        &deploy.UID,
+	}
+	// TODO(rz): Use ingress trait instead
+	// set up a service for the deployment if one of the containers have port
+	serviceCreated := false
+	var service *corev1.Service
+	for _, c := range deploy.Spec.Template.Spec.Containers {
+		// TODO (rz): pass in all the ports
+		if len(c.Ports) != 0 {
```

Comment:
maybe we need to create multiple services for ports

Reply:
good point, that's the part we actually want to discuss